pkgs/build-support/docker/default.nix

```diff
@@ -866,4 +869,96 @@ rec {
+
+      # A mapping from output name to the nix store path where they should end up
+      outputPaths = lib.genAttrs drv.outputs (output: builtins.unsafeDiscardStringContext drv.${output}.outPath);
+
+      # Additional binaries available in this image, both as the initial
+      # command of the arguments, and within a bash
+      binaryPath = lib.makeBinPath [ bashInteractive builder coreutils ];
+
+      setupPackage = runCommandNoCC "setup" {
+        # Mimics a standard nix-shell
+        rc = ''
+          source $stdenv/setup
+          set +e
+          runHook shellHook
+          export PATH=$PATH:${binaryPath}
+        '';
+        passAsFile = [ "rc" ];
+      } ''
+        mkdir -p $out/{etc/ssl/certs,tmp,bin,usr/bin,home/user/{outputs,build}}
+        ln -s ${bashInteractive}/bin/sh $out/bin/sh
+        ln -s ${coreutils}/bin/env $out/usr/bin/env
+        cp "$rcPath" $out/home/user/.bashrc
+        ln -s "${cacert}/etc/ssl/certs/ca-bundle.crt" $out/etc/ssl/certs/ca-certificates.crt
+
+        ${lib.concatStringsSep "\n" (lib.mapAttrsToList (name: path: ''
+          ln -s ${path} $out/home/user/outputs/${name}
+        '') outputPaths)}
+      '';
+
+      # Support for passAsFile
+      passAsFiles = lib.listToAttrs (map (attr:
+        lib.nameValuePair "${attr}Path" (writeText "pass-as-text-${attr}" drv.drvAttrs.${attr})
+      ) (drv.drvAttrs.passAsFile or []));
+
+      # We can't just use `toString` on all derivation attributes because that
+      # would not put path literals in the closure. So we explicitly copy
+      # those into the store here
+      stringValue = value: if builtins.typeOf value == "path" then "${value}"
+        else if builtins.isList value then toString (map stringValue value)
+        else toString value;
+
+      # Environment variables set in the image
+      envVars = lib.mapAttrs (name: stringValue) drv.drvAttrs // outputPaths // passAsFiles // {
+        # These are needed for stdenv setup to work
+        NIX_STORE = storeDir;
+        NIX_BUILD_TOP = "/home/user/build";
+
+        HOME = "/home/user";
+        TMPDIR = "/tmp";
+        PATH = binaryPath;
+      };
+
+      # A binary that calls the command to build the derivation
+      builder = writeShellScriptBin "buildDerivation" ''
+        exec ${lib.escapeShellArg (stringValue drv.drvAttrs.builder)} ${lib.escapeShellArgs (map stringValue drv.drvAttrs.args)}
+      '';
+
+    in buildLayeredImage {
+      inherit name tag;
+      contents = [ setupPackage ] ++ contents;
+      extraCommands = ''
+        # Creating nix/store here allows the user to write to it
+        # Which is necessary for the derivation output paths
+        mkdir -p ${lib.removePrefix "/" storeDir}
+      '';
+
+      # Run this image as the given uid/gid
+      config.User = "${toString uid}:${toString gid}";
+      # Set the owner of the files to that user as well
+      inherit uid gid;
+
+      # By default just starts a shell
+      config.Cmd = [ "bash" ];
```

Comment:
This should probably be easily overridable. I want to override it in my use case as I don't want to only be able to run containers interactively. Using `shellHook` to do run a task means it only works when `.bashrc` is loaded which means it must be interactive.